File: HW/HW10/levelOrderTraversal.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def levelOrder(root):
    if root is None:
        return []
    
    queue = []
    queue.append(root)
    finalList = []

    while queue:
        level = []
        size = len(queue)
        
        for count in range(size):
            curr = queue.pop(0)
            level.append(curr.val)
            
            if curr.left:
                queue.append(curr.left)
            if curr.right:
                queue.append(curr.right)
        
        finalList.append(level)
    
    return finalList

# root = TreeNode(4)
# root.left = TreeNode(3)
# root.right = TreeNode(8)
# root.right.right = TreeNode(9)
# root.right.left = TreeNode(5)
# root.left.left = TreeNode(1)

# print(levelOrder(root))

File: HW/HW10/LevelOrderTraversal.java
import java.util.*;

public class LevelOrderTraversal {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> finalList = new ArrayList<>();
        if (root == null) {
            return finalList;
        }
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        queue.add(root);

        while (!queue.isEmpty()) {    
            List<Integer> level = new ArrayList<Integer>();
            int size = queue.size();

            for (int i = 0; i < size; i++) {
                TreeNode curr = queue.poll();
                level.add(curr.val);

                if (curr.left != null)
                    queue.add(curr.left);
                if (curr.right != null)
                    queue.add(curr.right);
            }
            finalList.add(level);
        }
        return finalList;
     }


     public static void main(String[] args) {
        // TreeNode root = new TreeNode(4);
        // root.left = new TreeNode(3);
        // root.right = new TreeNode(8);
        // root.left.left = new TreeNode(1);
        // root.right.left = new TreeNode(5);
        // root.right.right = new TreeNode(9);

        // LevelOrderTraversal l = new LevelOrderTraversal();
        // System.out.println(l.levelOrder(root));
     }
 
}
